### PR TITLE
[SPARK-41238][CONNECT][PYTHON] Support more built-in datatypes

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, in the sever side, make `proto_datatype` <-> `catalyst_datatype` conversion support all the built-in sql datatypes;

2, in the client side, make `proto_datatype` <-> `pyspark_catalyst_datatype` conversion support [all the datatypes that are supported in pyspark now.](https://github.com/apache/spark/blob/master/python/pyspark/sql/types.py#L60-L83)

### Why are the changes needed?
right now, only `long`, `string`, `struct` are supported

```
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Does not support convert float to connect proto types."
        debug_error_string = "{"created":"@1669206685.760099000","description":"Error received from peer ipv6:[::1]:15002","file":"src/core/lib/surface/call.cc","file_line":1064,"grpc_message":"Does not support convert float to connect proto types.","grpc_status":2}"
```

this PR make the schema and literal expr support more datatypes.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
added UT

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]
This is nice! LGTM, pending tests.

We should also probably fix up literal support together after this (see also https://github.com/apache/spark/pull/38462#discussion_r1013933320)

**Comment by amaliujia** [comment]
Thanks. 

Can you test both `nullable=true` and `nullable=false` case?

**Comment by zhengruifeng** [comment]
> Can you test both nullable=true and nullable=false case?

the tests added in `test_schema` covers those cases

**Comment by grundprinzip** [comment]
Thank you! This has caused some challenges in demoing as we didn't support many types!

File: connector/connect/src/main/protobuf/spark/connect/types.proto [hunk 1]
```diff
@@ -26,31 +26,46 @@ option java_package = "org.apache.spark.connect.proto";
 // itself but only describes it.
 message DataType {
```
Comment (grundprinzip) [inline connector/connect/src/main/protobuf/spark/connect/types.proto:27]: This is a severe breaking change, we need to make sure that at some point in time our protos become stable. In particular, here the order of the members changes.

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 2]
```diff
@@ -144,6 +144,73 @@ def test_schema(self):
             schema,
         )
 
+        # test FloatType, DoubleType, DecimalType, StringType, BooleanType, NullType
+        query = """
```
Comment (grundprinzip) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:148]: Love the SQL test for type checking :)

**Review by grundprinzip — APPROVED** [review]

**Comment by amaliujia** [comment]
LGTM

**Comment by zhengruifeng** [comment]
merged into master

File: connector/connect/src/main/protobuf/spark/connect/types.proto (hunk 1 above)
Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/types.proto:27]: maybe also need to add some UDTs(like VectorUDT) in the future, but now SQL types were completed.

**Comment by zhengruifeng** [comment]
> We should also probably fix up literal support together after this (see also https://github.com/apache/spark/pull/38462#discussion_r1013933320)

will take a look

**Pull request closed without merging**